### PR TITLE
Add `childstate_getStorageEntries` RPC

### LINKED ISSUE #9203: Allow for childstate queries for multiple child keys
Current we have childstate queries for single keys, 

```
getStorage(
  childKey: PrefixedStorageKey,
  key: StorageKey,
  at: Option<Hash>
): Option<StorageData>
``` 

This approach work quite well. However in places where we do use childstate (e.g. crowdloans) we actually want to get multiple entries. For instance, I have 4 accounts for contributed to a specific crowdloan and to retrieve the values need to do 4 individual requests. This is in the cases where I know that the specific accounts have contributed.

Now look at the case where I may want to check if some of my accounts contributed to a specific loan. To do this would need to make X indv. RPC calls (where X = number of available accounts). For an RPC-user perspective and from an Node overhead perspective, it would potentially be much more optimal to have a signature such as -

```
/// getStorages? getEntries? Not sure about the naming...
getEntries(
  childKey: PrefixedStorageKey,
  keys: Vec<StorageKey>,
  at: Option<Hash>
): Vec<Option<StorageData>>
```

Which would allow us to query a specific childtrie to retrieve the values for more than one key in a single request.

### PR DESCRIPTION
**Pull request by hirschenberger**
Maybe it would make sense to deprecate `getStorage` as it can be substituted with `getStorageEntries` now?

fixes #9203

### CODE REVIEW

File: client/rpc/src/state/state_full.rs [hunk 1]
```diff
@@ -358,6 +361,22 @@ where
...
+	) -> FutureResult<Vec<Option<StorageData>>> {
+		let block = match self.block_or_best(block) {
+			Ok(b) => b,
+			Err(e) => return Box::new(result(Err(client_err(e)))),
+		};
+		let client = self.client.clone();
+		Box::new(join_all(
+			keys.into_iter()
+				.map(move |key| client.storage(&BlockId::Hash(block), &key).map_err(client_err)),
+		))
```
Comment (bkchr) [inline client/rpc/src/state/state_full.rs:377]:
```suggestion
		Box::new(keys.into_iter()
				.map(move |key| client.storage(&BlockId::Hash(block), &key).map_err(client_err))
				.collect::<Result<_, _>>()
		)
```

There are no futures involved.

File: client/rpc/src/state/state_full.rs [hunk 2]
```diff
@@ -722,6 +741,30 @@ where
...
+		storage_key: PrefixedStorageKey,
+		keys: Vec<StorageKey>,
+	) -> FutureResult<Vec<Option<StorageData>>> {
+		let block = match self.block_or_best(block) {
+			Ok(b) => b,
+			Err(e) => return Box::new(result(Err(client_err(e)))),
+		};
+		let client = self.client.clone();
+		Box::new(
+			join_all(keys.into_iter().map(move |key| {
```
Comment (bkchr) [inline client/rpc/src/state/state_full.rs:756]: Same here

File: client/rpc-api/src/child_state/mod.rs [hunk 3]
```diff
@@ -66,6 +66,15 @@ pub trait ChildStateApi<Hash> {
 		hash: Option<Hash>,
 	) -> FutureResult<Option<StorageData>>;
 
+	/// Returns child storage entries for multiple keys at a specific block's state.
+	#[rpc(name = "childstate_getStorages")]
+	fn storages(
```
Comment (bkchr) [inline client/rpc-api/src/child_state/mod.rs:71]: `storage_entries` maybe? `storages` sounds weird.

File: client/rpc/src/state/state_light.rs [hunk 4]
```diff
@@ -245,6 +245,26 @@ where
...
+	fn storages(
+		&self,
+		block: Option<Block::Hash>,
+		keys: Vec<StorageKey>,
+	) -> FutureResult<Vec<Option<StorageData>>> {
+		let remote_blockchain = self.remote_blockchain.clone();
+		let fetcher = self.fetcher.clone();
+		let block = self.block_or_best(block);
+		Box::new(join_all(keys.into_iter().map(move |key| {
+			storage(&*remote_blockchain, fetcher.clone(), block, vec![key.0.clone()])
```
Comment (bkchr) [inline client/rpc/src/state/state_light.rs:257]: You should request all keys in one request and not one request per key.

File: client/rpc/src/state/state_light.rs [hunk 5]
```diff
@@ -245,6 +245,26 @@ where
...
+				.boxed()
+				.compat()
+				.map(move |mut values| {
+					values
+						.remove(&key)
+						.expect("successful request has entries for all requested keys; qed")
```
Comment (bkchr) [inline client/rpc/src/state/state_light.rs:263]: Are you sure?

File: client/rpc/src/state/state_light.rs [hunk 6]
```diff
@@ -563,6 +583,52 @@ where
...
+			let child_storage =
+				resolve_header(&*remote_blockchain, &*fetcher, block).then(move |result| {
+					match result {
+						Ok(header) => Either::Left(
+							fetcher2
+								.remote_read_child(RemoteReadChildRequest {
+									block,
+									header,
+									storage_key,
+									keys: vec![key.0.clone()],
```
Comment (bkchr) [inline client/rpc/src/state/state_light.rs:607]: Again you are only requesting one key after another.

**Review by hirschenberger — COMMENTED** [review]
followup PR coming...

**Comment by hirschenberger** [comment]
@bkchr Can I have another review?

File: client/rpc/src/state/state_full.rs [hunk 7]
```diff
@@ -356,6 +360,24 @@ where
...
 	}
 
+	fn storage_entries(
+		&self,
+		block: Option<Block::Hash>,
+		keys: Vec<StorageKey>,
+	) -> FutureResult<Vec<Option<StorageData>>> {
+		let block = match self.block_or_best(block) {
+			Ok(b) => b,
+			Err(e) => return Box::pin(err(client_err(e))),
```
Comment (bkchr) [inline client/rpc/src/state/state_full.rs:370]:
```suggestion
			Err(e) => return err(client_err(e)).boxed(),
```

File: client/rpc/src/state/state_full.rs [hunk 8]
```diff
@@ -356,6 +360,24 @@ where
...
+		};
+		let client = self.client.clone();
+		try_join_all(keys.into_iter().map(move |key| {
+			let client = client.clone();
+
+			async move { client.clone().storage(&BlockId::Hash(block), &key).map_err(client_err) }
```
Comment (bkchr) [inline client/rpc/src/state/state_full.rs:376]:
```suggestion
			let res = client.storage(&BlockId::Hash(block), &key).map_err(client_err);

			async move { res }
```

File: client/rpc/src/state/mod.rs [hunk 9]
```diff
@@ -94,6 +94,13 @@ where
 		key: StorageKey,
 	) -> FutureResult<Option<StorageData>>;
 
+	/// Returns storage entries for multiple keys at a specific block's state.
+	fn storage_entries(
+		&self,
+		block: Option<Block::Hash>,
+		keys: Vec<StorageKey>,
+	) -> FutureResult<Vec<Option<StorageData>>>;
+
```
Comment (bkchr) [inline client/rpc/src/state/mod.rs:103]: For what do we need this function?

File: client/rpc/src/state/state_full.rs [hunk 10]
```diff
@@ -715,6 +737,35 @@ where
...
+	fn storage_entries(
+		&self,
+		block: Option<Block::Hash>,
+		storage_key: PrefixedStorageKey,
+		keys: Vec<StorageKey>,
+	) -> FutureResult<Vec<Option<StorageData>>> {
+		let child_info = match ChildType::from_prefixed_key(&storage_key) {
+			Some((ChildType::ParentKeyId, storage_key)) =>
+				Arc::new(ChildInfo::new_default(storage_key)),
+			None => return Box::pin(err(client_err(sp_blockchain::Error::InvalidChildStorageKey))),
```
Comment (bkchr) [inline client/rpc/src/state/state_full.rs:749]:
```suggestion
			None => return err(client_err(sp_blockchain::Error::InvalidChildStorageKey)).boxed(),
```

File: client/rpc/src/state/state_full.rs [hunk 11]
```diff
@@ -715,6 +737,35 @@ where
...
+		};
+		let block = match self.block_or_best(block) {
+			Ok(b) => b,
+			Err(e) => return Box::pin(err(client_err(e))),
```
Comment (bkchr) [inline client/rpc/src/state/state_full.rs:753]:
```suggestion
			Err(e) => return err(client_err(e)).boxed(),
```

File: client/rpc/src/state/state_full.rs [hunk 12]
```diff
@@ -715,6 +737,35 @@ where
...
+		let client = self.client.clone();
+		try_join_all(keys.into_iter().map(move |key| {
+			let client = client.clone();
+			let child_info = child_info.clone();
+			async move {
+				client
+					.clone()
+					.child_storage(&BlockId::Hash(block), &child_info, &key)
+					.map_err(client_err)
+			}
```
Comment (bkchr) [inline client/rpc/src/state/state_full.rs:764]:
```suggestion
			let res = client
					.clone()
					.child_storage(&BlockId::Hash(block), &child_info, &key)
					.map_err(client_err);
					
			async move {
				res
			}
```

File: client/rpc/src/state/state_light.rs [hunk 13]
```diff
@@ -531,6 +542,56 @@ where
...
+												.filter_map(|(k, d)| {
+													if keys.contains(k) {
+														if let Some(d) = d {
+															return Some(Some(StorageData(
+																d.to_vec(),
+															)))
+														}
+													};
+
+													None
```
Comment (bkchr) [inline client/rpc/src/state/state_light.rs:580]:
```suggestion
													keys.contains(k).then(|| d.map(|v| StorageData(v.to_vec())))
```

File: client/rpc/src/state/state_light.rs [hunk 14]
```diff
@@ -531,6 +542,56 @@ where
...
+										})
+										.map_err(client_err),
+								)
+							}),
+					),
+					Err(error) => Either::Right(ready(Err(error))),
+				}
+			});
+
+		Box::pin(child_storage.boxed())
```
Comment (bkchr) [inline client/rpc/src/state/state_light.rs:592]:
```suggestion
		child_storage.boxed()
```

**Review by bkchr — COMMENTED** [review]
Sorry for the delay. Some last nitpicks, then we are good to merge this

File: client/rpc/src/state/mod.rs (hunk 9 above)
Comment (hirschenberger) [reply client/rpc/src/state/mod.rs:103]: bc `StateApi` impl for `State` requires it.

Comment (bkchr) [reply client/rpc/src/state/mod.rs:103]: Can you please link me to what this requires?

Comment (hirschenberger) [reply client/rpc/src/state/mod.rs:103]: https://github.com/paritytech/substrate/blob/267dac2afdef51554eeb881fac0d844dd2761c92/client/rpc/src/state/mod.rs#L266

Comment (bkchr) [reply client/rpc/src/state/mod.rs:103]: And where does it need this function in the implementation there?

Comment (hirschenberger) [reply client/rpc/src/state/mod.rs:103]:
```rust
error[E0046]: not all trait items implemented, missing: `storage_keys`
   --> client/rpc/src/state/mod.rs:273:1
    |
273 | / impl<Block, Client> StateApi<Block::Hash> for State<Block, Client>
274 | | where
275 | |     Block: BlockT + 'static,
276 | |     Client: Send + Sync + 'static,
...   |
431 | |     }
432 | | }
    | |_^ missing `storage_keys` in implementation
    |
    = help: implement the missing item: `fn storage_keys(&self, _: StorageKey, _: std::option::Option<Hash>) -> Pin<Box<(dyn futures::Future<Output = Result<Vec<StorageKey>, sc_rpc_api::state::error::Error>> + std::marker::Send + 'static)>> { todo!() }`
    ```

Comment (hirschenberger) [reply client/rpc/src/state/mod.rs:103]: Sorry, it tracked the wrong function, forget the above comments.

Comment (hirschenberger) [reply client/rpc/src/state/mod.rs:103]: The function is not required for the RPC call. 
I implemented them to have the same set of query possibilities, if you say we don't need them I can remove them.

Comment (bkchr) [reply client/rpc/src/state/mod.rs:103]: Yeah please remove it.

**Comment by hirschenberger** [comment]
Should be ready now.

**Comment by bkchr** [comment]
@hirschenberger please run `cargo fmt`

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Comment by hirschenberger** [comment]
@bkchr There's some CI issue here?

**Comment by bkchr** [comment]
bot merge

**Pull request merged**